### DISCUSSION TITLE
Windows: SO CLOSE (I think!)

### DISCUSSION THREAD

**Discussion by ghost** [actions/actions-runner-controller#3037] [Questions]
Hey all!  

I've been through miles of forum posts as well as the official instructions for windows runners in the ARC docs, enabling Windows in my existing linux cluster.. and feel I am super close to getting a runner up.  Only place I'm stuck now is that when the windows runner comes up, I get this in the logs, then it dies..
```
--------------------------------------------------------------------------------
|        ____ _ _   _   _       _          _        _   _                      |
|       / ___(_) |_| | | |_   _| |__      / \   ___| |_(_) ___  _ __  ___      |
|      | |  _| | __| |_| | | | | '_ \    / _ \ / __| __| |/ _ \| '_ \/ __|     |
|      | |_| | | |_|  _  | |_| | |_) |  / ___ \ (__| |_| | (_) | | | \__ \     |
|       \____|_|\__|_| |_|\__,_|_.__/  /_/   \_\___|\__|_|\___/|_| |_|___/     |
|                                                                              |
|                       Self-hosted runner registration                        |
|                                                                              |
--------------------------------------------------------------------------------

# Authentication

No such host is known. (api.github.com:443)
```

kubernetes version: 1.23
runner deployment: 
```
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: my-repo-action-runner-deploy-win
  namespace: actions-runners
spec:
  template:
    spec:
      image: public.ecr.aws/e7b3d8i9/windows-2019-arc-runner:latest
      nodeSelector:
        kubernetes.io/os: windows
        kubernetes.io/arch: amd64
      dockerEnabled: false
      tolerations:
        - key: "os"
          operator: "Equal"
          value: "windows"
          effect: "NoSchedule"
      labels:
        - windows
        - X64
      repository: my/my-repo
      resources:
        limits:
          cpu: "8.0"
          memory: "16Gi"
        requests:
          cpu: "8.0"
          memory: "16Gi"
```

note: 
- this runnerdeploy is a bit different than the one specified in the official docs [here](https://github.com/actions/actions-runner-controller/blob/master/docs/configuring-windows-runners.md) - but when I use the one in the official docs, I get the error "windows cannot run in privileged mode"  - i've added the taint to the pod to further try to ensure that a windows pod doesn't try to get scheduled on a linux node (per another windows issue I found in the issues list) - added the appropriate toleration to the karpenter node provisioner also.  
- the docker image i'm referencing was built from the exact dockerfile in the [offical docs](https://github.com/actions/actions-runner-controller/blob/master/docs/configuring-windows-runners.md)
- I followed the "enabling windows" instructions [here](https://docs.aws.amazon.com/eks/latest/userguide/windows-support.html)
- my arc version (installed via helm chart) 

```
NAME                            NAMESPACE               REVISION        UPDATED                                 STATUS          CHART                                   APP VERSION
actions-runner-controller       actions-runner-system   3               2023-10-20 15:52:31.530551 -0400 EDT    deployed        actions-runner-controller-0.23.5        0.27.5  
```

In short, I've _finally_ got it scheduling (got through the "no Privateipv4address" issue as well), and now it just seems like the docker container can't access the host's internet... maybe?  I mean, the image was pulled.. so the node must have connectivity...? 

If I can provde any additional info, please let me know!  Thank you in advance!  Been working on this on and off for 2 weeks :(

**Comment by adiroiban** [comment]
Thanks for sharing the progress.

Maybe I might be asking a stupid question.

Is your Windows k8s worker node working? Does it have internet access?

Have you tried deploying a single pod, based on `mcr.microsoft.com/windows/servercore:ltsc2019` and then attach to the pod and see if you have internet access there?

Regards

**Reply by ghost** [reply]
Yep you're right. I ran a pod with this based on that image
```
          image: mcr.microsoft.com/windows/servercore:ltsc2019-amd64
          ports:
            - name: http
              containerPort: 80
          imagePullPolicy: IfNotPresent
          command:
            - powershell.exe
            - -command
            - "ping 8.8.8.8"
```

.. and nope! `request timed out`.  Now to see why.... :/ 

the nodegroup was configured by eksctl and my linux nodes in the same cluster are fine, fwiw... will investigate!

**Reply by adiroiban** [reply]
The container images is pulled by the host node, and I think that the image pull operation happens outside of the k8s/pod network environment.

**Reply by adiroiban** [reply]
Have you fixed the issue?
I'm now giving a try to ARC on Windows.
I have started with Calico and I got a similar issue. DNS is not setup for pod and NAT is not working.

**Reply by ghost** [reply]
No friend, I am trying to figure out why the (container? host?) doesn't have internet access - am not 100% on it at the moment but if i find something I will share it!  Disappointed b/c I set it up with eksctl.  Kinda feel like it's an issue btw docker and the host and not the host/instance/node itself but will keep checking.

**Reply by ghost** [reply]
Just connected to the windows instance and verified that the instance itself has internet access but the docker container does not.  Zeroing in...

**Reply by adiroiban** [reply]
I got the network part fixed for my k8s cluster on Windows. It was a `containerd` misconfiguration.

**Reply by ghost** [reply]
Can you share that solution with me @adiroiban ?  I've found and tried a couple to no avail. Thinking I might have to delete then reprovision the nodegroup with eksctl

**Reply by adiroiban** [reply]
I am not using AWS. I have a bare metal node with Calico.
I have installed containerd and then https://docs.tigera.io/calico/latest/getting-started/kubernetes/windows-calico/quickstart

I don't use microk8s but basically I followed the same step.
https://microk8s.io/docs/add-a-windows-worker-node-to-microk8s

**Reply by ghost** [reply]
Narrowing it further, it seems nodes provisioned via eksctl in a nodegroup can access the internet just fine, the runner starts and waits just fine... so that's good.   It seems to be the nodes I'm having _karpenter_ bring up have the connectivity issue.  I've compared the instances in AWS side by side and most everything is the same except for a few extra secondary IP addresses being assigned to the (working) nodegroup windows node... still digging  :D

.. noting of course that my karpenter linux nodes can access the internet just fine.. so something related to the karpenter + windows combination..

**Reply by ghost** [reply]
Got iiiiit!   Had to specify a different DNS server for the karpenter windows provisioner and so far all seems to be good -  https://github.com/aws/karpenter/issues/4420#issuecomment-1800902865

**Comment by adiroiban** [comment]
Here is my config for the helm deployment of the newer ARC controller

https://gist.github.com/adiroiban/cd3ccdb9b28e949123bb51914de8d96b

It looks like ARC listeners is happy to spin any container and inject the JIT token.

**Reply by ghost** [reply]
Appreciate this!  My latest road block is the image I built using the ARC dockerfile doesn’t have visual studio on it, which I need for unity builds… but that doesn’t have anything to do with ARC.. I just need an image with VS on it or need to install it in the dockerfile..

**Reply by adiroiban** [reply]
Have you checked the scripts from here https://github.com/actions/runner-images/blob/main/images/win/scripts/Installers/Install-VS.ps1

With servercore you should be able to run everything that is in there.

-------

My project is very simple and I was trying to get nanoserver up and running, but `actions/checkout` is not happy with mingit

**Comment by quoctruong** [comment]
@adiroiban, @gotchipete Were you able to enable Kubernetes Mode as well for the Windows Container? Or is this not supported?

**Reply by adiroiban** [reply]
Yes. The ARC helm chart works well with Kubernetes on Windows.

Have you checked my previous gist ?

Here is the updated gist with an example config https://gist.github.com/adiroiban/a3a4444d77d68e372d5578e31548a1f0

**Reply by quoctruong** [reply]
Thanks @adiroiban. I meant this mode (https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/deploying-runner-scale-sets-with-actions-runner-controller#using-docker-in-docker-or-kubernetes-mode-for-containers) so a different container will be spun up for the workload. I was trying that out just on a normal Windows Container and I ran into
```
Error: Container operations are only supported on Linux runners
```

**Reply by adiroiban** [reply]
Ah...the service containers and run step in container. No. I have not tried that with Windows.

That works great with Ubuntu.


 Is that functionality supposed with GitHub own VMs and runners?

**Reply by quoctruong** [reply]
I was looking at some issues like https://github.com/actions/runner/issues/904 which seems to indicate that it is not supported. But at least the normal Windows flow seem to work based on the gist you shared.

**Reply by adiroiban** [reply]
As far as I know [Running jobs in a container](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container) and [Service contaienrs](https://docs.github.com/en/actions/using-containerized-services/about-service-containers) are only supported on Ubuntu ... not even macOS.

There is no mention about Ubuntu only support in the docs.

---------

Docker-in-docker and kubernetes pods, all this code is open source, so in theory you can look to implement the required support for Windows

---------

But starting a simple Windows pod with a just in time / emphemeral GitHub Action runner is supported on Windows.

You will need to build your Windows image and define a separate scaleset ... but it works just fine

```
      - name: runner
        image: ghcr.io/YOUR-GITHUB-ORG/arc-runner-windows:2.313.0.x64.1
        command: ["run.cmd"]
        args: ["--jitconfig", "$(ACTIONS_RUNNER_INPUT_JITCONFIG)"]
```

**Reply by adiroiban** [reply]
The original question is not about docker-in-docker or container support for Windows.

Also, the original question is about the legacy `apiVersion: actions.summerwind.dev/v1alpha1` with support from the community.

I am using the newer API supported by the GitHub company/team and installed using `helm`

**Reply by denny-snapone** [reply]
Hi @adiroiban, 

Is it the newest Runner Scale Set?
